### PR TITLE
Add org-level workspace class restrictions

### PR DESCRIPTION
**Pull request by mustard-mh**
## Description

<details>
<summary>Summary generated by Copilot</summary>

### <samp>🤖[[deprecated]](https://githubnext.com/copilot-for-prs-sunset) Generated by Copilot at 6f2434d</samp>

Added a new organization setting to restrict the workspace class of the organization members. Implemented the backend logic, database schema, public API, and UI components to support this feature. Updated the tests and cache version accordingly.

</details>

## Related Issue(s)

Fixes EXP-1010, EXP-1053

## How to test

- Update max workspace class in org settings page
- Create workspaces with ws classes out of scopes should be failed
- Create workspaces with ws classes in scopes should work
- Join this org, we should not able to create workspaces, as its allowed workspace classes are not available in installation
  - https://hw-org-cls.preview.gitpod-dev.com/orgs/join?inviteId=ef169233-0591-4dbb-9901-58f7ca07b3d7
  - If invited org setting is correct (has something selected) again, you can do it manually with your org with commands below

```bash
# Terminal #1
kubectl port-forward svc/mysql 3307:3306

# Terminal #2
kubectl get secrets/mysql -ojson | jq -r .data.password | base64 -d

mysql -h127.0.0.1 --port 3307 -p

# inside mysql
USE gitpod;
UPDATE d_b_org_settings SET allowedWorkspaceClasses='[\"something-not-defined\"]' WHERE orgId='<your_setting_id>';
```

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - hw-org-cls</li>
	<li><b>🔗 URL</b> - <a href="https://hw-org-cls.preview.gitpod-dev.com/workspaces" target="_blank">hw-org-cls.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
	<li><b>📦 Version</b> - hw-org-cls-gha.21303</li>
	<li><b>🗒️ Logs</b> - <a href="https://console.cloud.google.com/logs/query;query=jsonPayload.kubernetes.host%3D%22preview-hw-org-cls%22%0A%0A--%20Filter%20on%20service:%0A--%20jsonPayload.serviceContext.service%3D%22ws-manager-mk2%22%0A;duration=P1D?project=gitpod-core-dev" target="_blank">GCP Logs Explorer</a></li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [x] /werft preemptible
      Saves cost. Untick this only if you're really sure you need a non-preemtible machine.
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

File: components/dashboard/src/teams/TeamSettings.tsx [hunk 1]
```diff
@@ -397,3 +409,79 @@ function OrgDefaultWorkspaceImageModal(props: OrgDefaultWorkspaceImageModalProps
...
+        async (cls: string) => {
+            try {
+                await updateTeamSettings.mutateAsync(
+                    {
+                        ...props.settings,
+                        maximumWorkspaceClass: cls,
+                    },
+                    {
+                        onSuccess: () => {
+                            toast({ message: "Organization workspace size updated" });
```
Comment (Siddhant-K-code) [inline components/dashboard/src/teams/TeamSettings.tsx:434]: Current version: 

image


```suggestion
                            toast({ message: "Workspace size limit updated for your organization." });
```

File: components/dashboard/src/teams/TeamSettings.tsx [hunk 2]
```diff
@@ -249,6 +254,13 @@ function OrgSettingsForm(props: { org?: Organization; isOwner: boolean }) {
                     onClose={() => setShowImageEditModal(false)}
                 />
             )}
+
+            <Heading2 className="pt-12">Maximum Workspace Size</Heading2>
+            <Subheading className="max-w-2xl">
+                Limit the maximum workspace size available in your organization.
+            </Subheading>
+
+            <WorkspaceSizeOptions settings={settings} />
```
Comment (Siddhant-K-code) [inline components/dashboard/src/teams/TeamSettings.tsx:263]: I think, this action should be accessible to owners only? We should disable this if a Member is coming to Organization settings (just like Workspace sharing settings)

```suggestion
            <WorkspaceSizeOptions settings={settings} disabled={!isOwner} />
```

File: components/public-api/gitpod/v1/organization.proto [hunk 3]
```diff
@@ -98,6 +99,9 @@ message UpdateOrganizationSettingsRequest {
 
   // pass empty string to reset to the installation default workspace image
   optional string default_workspace_image = 4;
+
+  // pass empty string to unset maximum_workspace_class
+  optional string maximum_workspace_class = 5;
```
Comment (akosyakov) [inline components/public-api/gitpod/v1/organization.proto:104]: that's questionable, is there well defined order between classes? Classes are delivered from the workspace cluster, should be order defined there that the engine team does not break it later? What happens if the new class is added? cc @loujaybee 

alternatives is to allow enable/disable workspace classes

File: components/dashboard/src/teams/TeamSettings.tsx [hunk 4]
```diff
@@ -397,3 +406,78 @@ function OrgDefaultWorkspaceImageModal(props: OrgDefaultWorkspaceImageModalProps
...
+                </Alert>
+            )}
+            <Label>
+                {classes.map((wsClass) => (
+                    <CheckboxInputField
+                        id={wsClass.id}
+                        label={wsClass.displayName}
+                        hint={wsClass.description}
+                        checked={selectedValue.includes(wsClass.id)}
+                        onChange={(checked) => {
```
Comment (filiptronicek) [inline components/dashboard/src/teams/TeamSettings.tsx:467]: It would be great to make this consistent with the Configuration detail pages and have a <kbd>Save</kbd> button here instead of autosaving. It may be a bit unexpected if the user is familiar with those UIs.

File: components/dashboard/src/teams/TeamSettings.tsx [hunk 5]
```diff
@@ -397,3 +406,78 @@ function OrgDefaultWorkspaceImageModal(props: OrgDefaultWorkspaceImageModalProps
...
+        async (classes: string[]) => {
+            try {
+                await updateTeamSettings.mutateAsync(
+                    {
+                        ...props.settings,
+                        allowedWorkspaceClasses: classes,
+                    },
+                    {
+                        onSuccess: () => {
+                            toast({ message: "Available workspace classes updated for your organization." });
```
[the remark this answers is not in the document]
Comment (mustard-mh) [reply components/dashboard/src/teams/TeamSettings.tsx:431]: I'd like to keep `Available`

File: components/dashboard/src/teams/TeamSettings.tsx (hunk 4 above)
Comment (mustard-mh) [reply components/dashboard/src/teams/TeamSettings.tsx:467]: Use a `save` button can also save traffic

cc @corygrunk

File: components/dashboard/src/teams/TeamSettings.tsx [hunk 6]
```diff
@@ -397,3 +406,78 @@ function OrgDefaultWorkspaceImageModal(props: OrgDefaultWorkspaceImageModalProps
...
+                    {errorMsg}
+                </Alert>
+            )}
+            <Label>
+                {classes.map((wsClass) => (
+                    <CheckboxInputField
+                        id={wsClass.id}
+                        label={wsClass.displayName}
+                        hint={wsClass.description}
+                        checked={selectedValue.includes(wsClass.id)}
```
[the remark this answers is not in the document]
Comment (mustard-mh) [reply components/dashboard/src/teams/TeamSettings.tsx]: I revert this changes, since I'm not sure what state we need to show when user unchecked all, 
- Should it change to check all?  -> It's strange because I know that I want to uncheck it
- Or keep unchecked all -> then why first view is all checked? Will it leads to misunderstanding that allow nothing?

Update description above to make them know that select nothing means allow all kind of workspace classes

[the remark this answers is not in the document]
Comment (filiptronicek) [reply components/dashboard/src/teams/TeamSettings.tsx]: @mustard-mh I think we should make it more explicit that you (probably) can't disable all classes. It feels like the pattern of "nothing is selected" = "everything is selected" is a bit flawed. I can see us starting off with everything checked (`.length === 0`) and then require that at least one checkbox remains checked at all times to avoid confusion.

[the remark this answers is not in the document]
Comment (mustard-mh) [reply components/dashboard/src/teams/TeamSettings.tsx]: @filiptronicek  Update to show message that there should have at least one class slected

image

File: components/dashboard/src/teams/TeamSettings.tsx [hunk 7]
```diff
@@ -397,3 +405,86 @@ function OrgDefaultWorkspaceImageModal(props: OrgDefaultWorkspaceImageModalProps
...
+                        ...props.settings,
+                        allowedWorkspaceClasses: classes,
+                    },
+                    {
+                        onSuccess: () => {
+                            toast({ message: "Available workspace classes updated." });
+                        },
+                    },
+                );
+            } catch (error) {
```
Comment (selfcontained) [inline components/dashboard/src/teams/TeamSettings.tsx:434]: Rather than catching this here and tracking our own `errorMsg` state we should use the `error` property that the `updateTeamSettings` hook result has, i.e. `updateTeamSettings.isError` && `updateTeamSettings.error`. This handles all of that state for us automatically.

File: components/dashboard/src/teams/TeamSettings.tsx [hunk 8]
```diff
@@ -397,3 +405,86 @@ function OrgDefaultWorkspaceImageModal(props: OrgDefaultWorkspaceImageModalProps
...
+                        const oldVal = selectedValue;
+                        let tmpSelectedValue = selectedValue;
+                        if (tmpSelectedValue.length === 0) {
+                            tmpSelectedValue = classes.map((e) => e.id);
+                        }
+                        const newVal = checked
+                            ? [...tmpSelectedValue, wsClass.id]
+                            : tmpSelectedValue.filter((e) => e !== wsClass.id);
+                        if (newVal.length === 0) {
+                            setErrorMsg("Should have one workspace class selected at least.");
```
Comment (selfcontained) [inline components/dashboard/src/teams/TeamSettings.tsx:476]: Rather than trying to track this state of if a class is selected we can derived the value and avoid state. For example, in our render method we can do something like:

```tsx
const noClassesSelected = selectedClasses.length === 0;
```

Then render an error message for that case. In regards to that, we can wrap the list of `<CheckboxInputFIeld>` components with `<CheckboxListField>` and that provides a property to render errors associated to those fields. For example:

```tsx
<CheckboxListField label="..." error="...">
  <CheckboxInputFIeld .../>
  <CheckboxInputFIeld .../>
</CheckboxListField>
```

File: components/dashboard/src/teams/TeamSettings.tsx [hunk 9]
```diff
@@ -397,3 +406,100 @@ function OrgDefaultWorkspaceImageModal(props: OrgDefaultWorkspaceImageModalProps
...
+                        checked={(!isChanged && noClassesSelected) || selectedValue.includes(wsClass.id)}
+                        onChange={(checked) => {
+                            const previousValue =
+                                !isChanged && noClassesSelected ? classes.map((e) => e.id) : selectedValue;
+                            setIsChanged(true);
+                            const newVal = checked
+                                ? [...previousValue, wsClass.id]
+                                : previousValue.filter((e) => e !== wsClass.id);
+                            setValidateError(
+                                newVal.length === 0 ? "Should have one workspace class selected at least." : "",
```
Comment (filiptronicek) [inline components/dashboard/src/teams/TeamSettings.tsx:480]:
```suggestion
                                newVal.length === 0 ? "At least one workspace class has to be selected." : "",
```

File: components/dashboard/src/teams/TeamSettings.tsx [hunk 10]
```diff
@@ -397,3 +406,100 @@ function OrgDefaultWorkspaceImageModal(props: OrgDefaultWorkspaceImageModalProps
...
+                        disabled={props.disabled || !isChanged || validateError.length > 0}
+                        loading={updateTeamSettings.isLoading}
+                        onClick={() => {
+                            handleUpdateTeamSettings(selectedValue);
+                        }}
+                    >
+                        Save
+                    </LoadingButton>
+                )}
+                {validateError.length > 0 && <span className="text-red-600 dark:text-red-400">{validateError}</span>}
```
Comment (filiptronicek) [inline components/dashboard/src/teams/TeamSettings.tsx:501]: Can we put this on a line before the button like we do with configs? 🙏 
image

**Comment by loujaybee** [comment]
Raised an issue regarding [the unset state of the select field for class](https://linear.app/gitpod/issue/EXP-1052/unset-state-on-workspace-class-toggle-misleading).

**Comment by mustard-mh** [comment]
Pushed a commit https://github.com/gitpod-io/gitpod/pull/19191/commits/8c4c73928cf05353af6dae900d895e985af7f552 for edge case: all selected allowed workspace classes are removed from installation.

Test case should be:
- Update DB to allow a known workspace class in org
- Start workspace should failed

Going to look into what dashboard UX will be

File: components/dashboard/src/components/SelectWorkspaceClassComponent.tsx [hunk 11]
```diff
@@ -39,6 +39,12 @@ export default function SelectWorkspaceClassComponent({
         if (!workspaceClasses) {
             return;
         }
+        if (workspaceClasses.length === 0) {
+            setError?.(
+                "Something unexpected has occurred, there's no allowed workspace class available in organization, please contact an admin.",
```
Comment (laushinka) [inline components/dashboard/src/components/SelectWorkspaceClassComponent.tsx:44]:
```suggestion
                "No allowed workspace classes available. Please contact your admin.",
```

File: components/server/src/workspace/workspace-service.ts [hunk 12]
```diff
@@ -93,25 +94,66 @@ export class WorkspaceService {
...
+            workspaceClassOverride,
+            this.config.workspaceClasses,
+        );
+        const settings = await this.orgService.getSettings(userId, organizationId);
+        if (settings.allowedWorkspaceClasses && settings.allowedWorkspaceClasses.length > 0) {
+            const hasOptions = await this.orgService.hasAllowedWorkspaceClassesInInstallation(userId, organizationId);
+            if (!hasOptions) {
+                throw new ApplicationError(
+                    ErrorCodes.PRECONDITION_FAILED,
+                    "Something unexpected has occurred, there's no allowed workspace class available in organization, please contact an admin",
```
Comment (laushinka) [inline components/server/src/workspace/workspace-service.ts:130]:
```suggestion
                    "No allowed workspace classes available. Please contact your admin.",
```

File: components/dashboard/src/data/organizations/orgs-query.ts [hunk 13]
```diff
@@ -14,12 +14,14 @@ import { Organization } from "@gitpod/public-api/lib/gitpod/v1/organization_pb";
...
 export function useOrganizationsInvalidator() {
     const user = useCurrentUser();
+    const orgs = useOrganizations();
 
     const queryClient = useQueryClient();
     return useCallback(() => {
         console.log("Invalidating orgs... " + JSON.stringify(getQueryKey(user?.id)));
         queryClient.invalidateQueries(getQueryKey(user?.id));
-    }, [user?.id, queryClient]);
+        orgs.refetch();
```
Comment (mustard-mh) [inline components/dashboard/src/data/organizations/orgs-query.ts:23]: Refetch it so that settings are up-to-date, if not, it will show old settings again, and needs another web fresh

File: components/dashboard/src/data/organizations/org-workspace-image-query.ts [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+            return settings.workspaceClasses || [];
+        },
+        {
+            enabled: !!organizationId,
+        },
+    );
+}
+
+function getQueryKey(organizationId?: string) {
+    return noPersistence(["listOrganizationWorkspaceClasses", organizationId || "undefined"]);
```
Comment (mustard-mh) [inline components/dashboard/src/data/organizations/org-workspace-image-query.ts:41]: Not cache workspace classes

Comment (filiptronicek) [reply components/dashboard/src/data/organizations/org-workspace-image-query.ts:41]: @mustard-mh why not? 👀

**Review by filiptronicek — APPROVED** [review]
LGTM
